src/gk.rs

```diff
@@ -45,21 +54,86 @@ type ConnectHTTPDefaultProvider = FillProvider<
 >;
 pub type GasKillerDefault = GasKiller<ConnectHTTPDefaultProvider>;
 
-pub struct GasKiller<P> {
-    _anvil: AnvilInstance,
-    provider: P,
-    code: Bytes,
+/// Builder for configuring and creating a `GasKiller` instance.
+pub struct GasKillerBuilder {
+    fork_url: Url,
+    block_number: Option<u64>,
+    timestamp: Option<u64>,
 }
 
-impl GasKiller<ConnectHTTPDefaultProvider> {
-    pub async fn new(fork_url: Url, block_number: Option<u64>) -> Result<Self> {
-        let anvil_init = Anvil::new().fork(fork_url.as_str());
+impl GasKillerBuilder {
+    /// Create a new builder with the required fork URL.
+    pub fn new(fork_url: Url) -> Self {
+        Self {
+            fork_url,
+            block_number: None,
+            timestamp: None,
+        }
+    }
+
+    /// Set the block number to fork from. If not set, forks from the latest block.
+    pub fn block_number(mut self, block_number: u64) -> Self {
+        self.block_number = Some(block_number);
+        self
+    }
+
+    /// Set the timestamp for the next block. If not set, defaults to `fork_block_timestamp + 1`.
+    /// The timestamp must be >= the fork block timestamp to satisfy Ethereum's strictly
+    /// increasing timestamp requirement.
+    pub fn timestamp(mut self, timestamp: u64) -> Self {
+        self.timestamp = Some(timestamp);
+        self
+    }
 
-        let anvil = if let Some(number) = block_number {
-            anvil_init.fork_block_number(number - 1).try_spawn()?
+    /// Build the `GasKiller` instance.
+    pub async fn build(self) -> Result<GasKillerDefault> {
+        // First, query the fork block's timestamp from the source RPC
+        // This ensures deterministic timestamps across all nodes
+        let source_provider = ProviderBuilder::new().connect_http(self.fork_url.clone());
+
+        let fork_block_timestamp = if let Some(number) = self.block_number {
+            let block = source_provider
+                .get_block_by_number(number.into())
+                .await?
+                .ok_or_else(|| anyhow!("Fork block {} not found", number))?;
+            block.header.timestamp
+        } else {
+            let block = source_provider
+                .get_block_by_number(alloy::eips::BlockNumberOrTag::Latest)
+                .await?
+                .ok_or_else(|| anyhow!("Latest block not found"))?;
+            block.header.timestamp
+        };
+
+        // Determine the timestamp to use for the next block
```

Comment:
what do you think the timestamp manipulations here are contributing?

Reply:
We emit a log when calling `sum` that includes the block timestamp https://github.com/BreadchainCoop/gas-killer-array-summation/blob/9ad682424ca12e4bc8f2819ab28a1d629213f147/src/ArraySummation.sol#L73